concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1161 @@
+
+<details>
+	<summary>classic way</summary>
+	
+```golang
+package main
+
+import "fmt"
+
+type foo struct {
+	name string
+	info string
+	bar  *bar
+}
+
+type bar struct {
+	name string
+}
+
+func NewFoo(name, info string, bar *bar) *foo {
+	return &foo{
+		name: name,
+		info: info,
+		bar:  bar,
+	}
+}
+
+func main() {
+
+	bar := &bar{"bar"}
+
+	foo := NewFoo("name",
+		"info",
+		bar,
+	)
+
+	fmt.Printf("%v", foo)
+}
+```
+</details>
+
+The pattern will change this code to
+<details>
+	<summary>With pattern</summary>
+	
+```golang
+package main
+
+import "fmt"
+
+type foo struct {
+	name string
+	info string
+	bar  *bar
+}
+
+type bar struct {
+	name string
+}
+
+type Opt func(f *foo)
+
+func WithInfo(info string) Opt {
+	return func(f *foo) {
+		f.info = info
+	}
+}
+
+func WithBar(bar *bar) Opt {
+	return func(f *foo) {
+		f.bar = bar
+	}
+}
+
+func NewFoo(name string, opts ...Opt) *foo {
+	foo := &foo{
+		name: name,
+	}
+	for _, opt := range opts {
+		opt(foo)
+	}
+	return foo
+}
+
+func main() {
+
+	bar := &bar{"bar"}
+
+	foo := NewFoo("name",
+		WithInfo("info"),
+		WithBar(bar),
+	)
+
+	fmt.Printf("%v", foo)
+
+}
+```
+	
+</details>
+
+**Advantage**:
+The benefit of this pattern is that with functional options it is easy to omit unnecessary arguments, e.g. omitting "info" in the classic way would require explicitly creating a new function to do this, or requires the user to explicitly pass "info" as an empty string. In the second implementation you just omit the call to `WithInfo`. 
+In the end this pattern can increase readability and allows reducing configurations to the really necessary information.
+
+The pattern works very well in tests where similar objects must be constructed multiple times. Here it is essential for the developer to quickly see what the important configuration of these objects for a given test case is.
+
+#### Guidelines for the pattern
+
+* Dont create Option functions for required arguments. As required arguments must not be omitted, it makes no sense to make them optional.
+* Names for Option functions should start with `With` followed by a short description of their purpose. Keep in mind that these names might be shortened by the developers IDE. Important information needs to be easily spotted in those names.
+e.g.
+
+```
+WithSinkFromSVC // configures a sink. The values are taken from a service
+WithProtocolBEB // configures the Protocol to be BEB
+```
+
+#### Bad Practice
+
+**Example 1**: Return function from WithSomething method
+<details>
+	<summary>Don't</summary>
+
+```go
+// source: https://github.com/kyma-project/kyma/pull/13242/files#diff-2f168bb71c1ca8d2f5781bac737393acd9bc3a1f829bdbe91093b7db54972433L130
+func WithExemptHandshakeBefore(p *eventingv1alpha1.ProtocolSettings) {
+	p.ExemptHandshake = func() *bool { // yes this could be refactored to utils.BoolPtr(true)
```

Comment:
can you make this a bit clearer? Is this about the BoolPtr or is it about not returning a ProtocolSettingsOpt from WithExemptHandshakeBefore?

Reply:
It is about `returning a ProtocolSettingsOpt from WithExemptHandshakeBefore`
I made it clearer. Please have another look 😃